core/lib/rom/schema.rb

```diff
@@ -416,6 +416,22 @@ def to_ast
       [:schema, [name, attributes.map(&:to_ast)]]
     end
 
+    def primary_key_name
+      if canonical?
+        @primary_key_name if defined? @primary_key_name
+      else
+        canonical.primary_key_name
+      end
+    end
+
+    def primary_key_names
+      if canonical?
+        @primary_key_names if defined? @primary_key_names
+      else
+        canonical.primary_key_name
+      end
+    end
```

Comment:
we shouldn't have this type of logic at run-time :/ How about we make these part of options instead? This way they can be maintained across instances.

Reply:
couldn't we just memoize them instead? Passing them as options is also not required, strictly speaking, because they are not always needed. And yes, either memoization or passing them with options is better than using ivars